commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/filesystem/FileSystemStorage.java

```diff
@@ -64,19 +72,33 @@ public InputStream fetch(final String key, final int from, final int to) throws
             throw new IllegalArgumentException("from cannot be more than to, from=" + from + ", to=" + to + " given");
         }
 
-        final Path path = Path.of(fsRoot, key);
+        final Path path = fsRoot.resolve(key);
         final long fileSize = Files.size(path);
         if (to > fileSize) {
             throw new IllegalArgumentException("to cannot be bigger than file size, to="
                 + to + ", file size=" + fileSize + " given");
         }
-        final InputStream result = new BoundedInputStream(Files.newInputStream(path), to + 1);
+
+        final int size = to + 1;
+        final InputStream result = new BoundedInputStream(Files.newInputStream(path), size);
         result.skip(from);
         return result;
     }
 
     @Override
     public void delete(final String key) throws IOException {
-        Files.delete(Path.of(fsRoot, key));
+        final Path path = fsRoot.resolve(key);
+        Files.deleteIfExists(path);
+        Path parent = path.getParent();
+        while (parent != null && Files.isDirectory(parent) && !parent.equals(fsRoot)) {
+            try (final DirectoryStream<Path> directory = Files.newDirectoryStream(parent)) {
+                if (!directory.iterator().hasNext()) {
+                    Files.deleteIfExists(parent);
+                } else {
+                    break;
+                }
+            }
+            parent = parent.getParent();
+        }
```

Comment:
I just realized that we have commons-io in the classpath already so maybe this? 
```suggestion
       while (parent != null && Files.isDirectory(parent) && !parent.equals(fsRoot)
            && PathUtils.isEmptyDirectory(parent)) {
            Files.deleteIfExists(parent);
            parent = parent.getParent();
        }
```

Reply:
Sure! Good catch